streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala

```diff
@@ -439,6 +465,30 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
         for (info <- receiverTrackingInfos.get(streamUID); eP <- info.endpoint) {
           eP.send(UpdateRateLimit(newRate))
         }
+      case ReceiverLaunchingTimeout(receiverId) => {
+        val timeoutFuture = receiverTimeoutFutures.remove(receiverId)
+        timeoutFuture.foreach { f =>
+          // If the receiver has not yet registered, start a new thread to stop StreamingContext
+          // gracefully.
+          new Thread("stopping-StreamingContext") {
+            setDaemon(true)
+
+            override def run(): Unit = {
+              if (isTrackerStarted) {
+                val stopSparkContext =
```

Comment:
Actually, you have to stop the context with an exception because something went wrong and the receiver could not be scheduled. Use reportError()